Apply changes for authentication endpoints

sunny-side-app opened the pull request:
PRが肥大化してしまい大変申し訳ないですが、認証系エンドポイントを作成しましたので、ご確認をお願いいたします。

## 使い方

- サーバーを起動させたらブラウザ上でURLにエンドポイントURL(例：http://127.0.0.1:8081/api/auth/register/ )を入力するとDRFのBrowsable API画面からデータを入力してリクエストを送信できる
- ターミナルからcurlなどで実行でもOK
    コマンド例(httpieを使った場合)：
    
`http POST http://127.0.0.1:8081/api/auth/register/ email="your_email@gmail.com" password="password_123" username="your_user_name"`

## 残対応
- ユーザーへ公開するログイン画面、サインアップ画面、メール認証結果画面の作成
- フィーチャーフラグの実装(他機能のテスト、CI/CD実行に影響しないようコード上でON/OFF切り替える)
- 認証機能の他機能への組み込み(組み込む対象機能の選定、組み込み方)
    -  対象の機能は下記で認識合っているか
        - role=adminのユーザーのみ可能
            - 管理ダッシュボードの参照/登録/編集
        - role=registeredのユーザーのみ可能
            - 購入履歴の参照
            - 個人プロフィールの参照/登録/編集
            - デジタル領収書の参照/登録/ダウンロード
    - 上記機能へどのように組み込むか

Commit:
Apply changes for authentication endpoints

haru864 reviewed (changes requested):
今回の修正目的に関わらないコードの差分がほとんどだと思うので、修正目的と変更内容が一致するようにしてほしいです。
また、コンフリクトしているので解消するようにお願いします。

sunny-side-app commented:
ご指摘ありがとうございます。また確認不足で大変失礼いたしました。
意図していない変更だったので確認して本日再度上げ直したいと思います。

Karukan0814 commented:
残対応についてですが、「ユーザーへ公開するログイン画面、サインアップ画面、メール認証結果画面の作成」→現状のフロントのヘッダー部分のユーザーマークを押下するとログイン画面立ち上がります。そこをキーにソースコード読んでバックエンドとつなげるといいです。

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,127 @@
+def send_confirmation_email(user):
+    uid = urlsafe_base64_encode(force_bytes(user.pk))
+    token = token_generator.make_token(user)
+    confirm_url = f'http://127.0.0.1:8081/api/auth/registration/account-confirm-email/{uid}/{token}/'
```
こういうURL系はenvファイルから読み込むようにしたほうがよい

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,127 @@
+            return JsonResponse({"message": "リフレッシュトークンが見つかりません。"}, status=400)
+
+# メール送信ロジック
+def send_confirmation_email(user):
```
APIView群の中にいきなり関係ないメソッドが出現するので、libフォルダ等に切り出して別管理したほうがよい

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,127 @@
+    token = token_generator.make_token(user)
+    confirm_url = f'http://127.0.0.1:8081/api/auth/registration/account-confirm-email/{uid}/{token}/'
+    
+    send_mail(
```
メール送信の失敗時に備えて　try-exceptでエラー時挙動を入れる

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,127 @@
+        send_reset_password_email(user)
+        return JsonResponse({"message": "パスワードリセットメールが送信されました。"}, status=200)
+
+def send_reset_password_email(user):
```
def send_confirmation_email(user):　と機能が重複している。処理を思いつくなら一つのメソッドにまとめたほうがよい。きれいにいかなそうならそのままでも。　　他指摘点はsend_confirmation_emailと同様

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,127 @@
+from django.conf import settings
+import json
+
+User = get_user_model()
```
ユーザー情報の登録先がアプリケーションclothes_shopであるのに、登録やログアウトをルートプロジェクトであるdjangopjのview.pyで行う意味とは何でしょうか。複数アプリケーションが存在し、認証機能を統一のものにする場合、ルートプロジェクトに認証機能を統一させるのは理にかなうと思います。しかし、ユーザー情報が個別アプリケーション固有のテーブルを使用するならば、認証機能は個別アプリケーションに落とすべきだと考えます。※現在動くなら即対応しなくていいかも

Karukan0814 commented on src/djangopj/urls.py:
```diff
@@ -1,7 +1,25 @@
 
 urlpatterns = [
-    path("admin/", admin.site.urls),
-    path("", include("clothes_shop.urls", namespace="clothes_shop")),
```
clothes_shopのHome画面に通じるパスが削除されていますが、これって、ログインしていないフラッとサイトに寄ったユーザーがどうHome画面に行くのでしょうか。clothes_shopの各エンドポイントで認証が必要なもの、必要でないものがあり、それぞれ認証系ミドルウェアのオンオフを切り替えます。動かしてはいないので申し訳ないのですが、このユーザー認証全体をみて、どのように個別アプリケーションclothes_shopに組み込むのか頭の中で私は整理できていません。

Karukan0814 reviewed (changes requested):
個別コメント参照だが、develpにマージする前にこの認証機能をどのようにclothes_shopのアプリケーションに組み入れるのか皆で確認した方がよい。確認点・修正点が多くなる気がします。

Commit:
Update .gitignore and resolve conflict in requirements.txt

Commit:
Fix .gitignore after rebase

Commit:
Address Karukan0814 comments on PR
- libフォルダに切り出し
- メール送信の失敗時に備えて　try-exceptでエラー時挙動を入れる
- send_confirmation_email, send_reset_password_email共通化
- urls.py修正

sunny-side-app commented:
@haru864 

下記対応させていただきました。
基本的にはdjangopj/views.py, djangopj/urls.pyが修正のメインとなり、Userモデルの修正とDockerやシェルスクリプト周りの一部修正(実行にあたり必要であった分)を修正しております。

> 今回の修正目的に関わらないコードの差分がほとんどだと思うので、修正目的と変更内容が一致するようにしてほしいです。
> また、コンフリクトしているので解消するようにお願いします。

@Karukan0814 

下記以外対応させていただきました。下記については、アプリclothes_shopとは別の共通的な機能、エンドポイントになるかと思い一旦別構成としていましたが、おっしゃる通り複数アプリにもならない気もしますので、ご記載の方針に修正でも問題ないと考えております。

> ユーザー情報の登録先がアプリケーションclothes_shopであるのに、登録やログアウトをルートプロジェクトであるdjangopjのview.pyで行う意味とは何でしょうか。複数アプリケーションが存在し、認証機能を統一のものにする場合、ルートプロジェクトに認証機能を統一させるのは理にかなうと思います。しかし、ユーザー情報が個別アプリケーション固有のテーブルを使用するならば、認証機能は個別アプリケーションに落とすべきだと考えます。※現在動くなら即対応しなくていいか

@haru864 @Karukan0814 
かるかんさんにコメントいただいたところでもあり、残対応にも記載しておりますが、そもそも想定されていた認証機能の想定と掛け離れていないか、既存機能にどのように組み込んでいくか認識合わせをさせていただきたいです。

**認証機能の他機能への組み込み(組み込む対象機能の選定、組み込み方)**
- 対象の機能は下記で認識合っているか
    - role=adminのユーザーのみ
        - 管理ダッシュボードの参照/登録/編集
    - role=registeredのユーザーのみ
        - 購入履歴の参照
        - 個人プロフィールの参照/登録/編集
        - デジタル領収書の参照/登録/ダウンロード
- 上記機能へどのように組み込むか

Karukan0814 commented:
以下、認証機能の組み込みについての私の認識です。
認証機能の他機能への組み込み(組み込む対象機能の選定、組み込み方)

role=adminのユーザーのみ
管理ダッシュボードの参照/登録/編集
role=registeredのユーザーのみ
購入履歴の参照
個人プロフィールの参照/登録/編集
デジタル領収書の参照/登録/ダウンロード
製品一覧と製品詳細画面からいいね・ウィッシュリスト操作（追加）
購入した製品についてのみ、購入履歴からレビュー・コメント（追加）

role=guestができること
製品一覧と製品詳細画面の閲覧
製品をカートに入れる→決済

Karukan0814 commented on src/clothes_shop/models.py:
```diff
@@ -55,20 +57,46 @@ class Product(models.Model):
+    is_staff = models.BooleanField(default=False) # Djangoの管理画面にアクセスするための特定の権限を制御(Django管理画面を利用する場合は必須)
+    first_name = models.CharField(max_length=30, blank=True)
+    last_name = models.CharField(max_length=30, blank=True)
+    profile_picture = models.ImageField(upload_to='profile_pictures/', null=True, blank=True)
```
画像を直接バックエンドサーバー上に入れる予定なのか？DBの容量圧迫するし、画像をフロントに一々送信することになる。ので、本番ではそうそう使わない技。定石としては画像をS3バケット等にいれ、そのURLパスをいれるようにする。admin画面でも画像は現在サーバー保存みたいなので、現状手をつけなくてもまあいいけれど。

Karukan0814 commented on src/clothes_shop/models.py:
```diff
@@ -55,20 +57,46 @@ class Product(models.Model):
+    email = models.EmailField(unique=True, null=True)
+    username = models.CharField(max_length=255, unique=True, null=True)
+    password = models.CharField(max_length=128, null=True)  # パスワードフィールド追加
+    role = models.CharField(max_length=50, choices=[('guest', 'Guest'), ('registered', 'Registered'), ('admin', 'Admin')],default='registered')  # guest, registered, admin
```
default='registered'にするのは危険球だを。処理忘れてゲストユーザーを登録ユーザーにしちゃう危険がでてくる。default='guest'にしといたほうがよい

Karukan0814 commented on src/djangopj/libs/email_utils.py:
```diff
@@ -0,0 +1,29 @@
+
+    try:
+        send_mail(subject, message, settings.DEFAULT_FROM_EMAIL, [user.email])
+    except Exception as e:
```
エラーでたときにPrintするのではなく、エラーログに書き出してほしい。

Karukan0814 commented:
コメント追加しました。おそらく追加ででてくるので、認識合わせ＋修正で時間を食うので、yadonさんはプロフィール画面に着手された方がよいかと思います。そのうえで、プロフィール画面を認証機能を使用してログインユーザーしか入れないようにするにはどうしたらよいかを考えると◎でないかな、と思います。

sunny-side-app commented:
@Karukan0814 
認証機能の具体的な組み込み方についてコメントいただきありがとうございます！
内容について承知しました。また、開発の進め方についても助言いただきありがとうございます。
以降は一旦プロフィール画面開発を進め、それを取っ掛かりとして組み込んでいきたいと思います。

sunny-side-app commented:
[PR#19認証についての補足.md](https://github.com/user-attachments/files/17529885/PR.19.md)
情報量が多すぎ＋説明不足のため簡単ですが補足コメントを用意しました。

sunny-side-app commented:
[PR#19補足.md](https://github.com/user-attachments/files/17529962/PR.19.md)
情報量が多い＋意図が説明不足だったため、補足コメントを用意しました。

Karukan0814 commented on mysql-variables.env:
```diff
@@ -0,0 +1,4 @@
+MYSQL_ROOT_PASSWORD=root_password
```
.envをコミットしておるぞ！！！！

Karukan0814 replied:
dockerignoreファイルをどう変更すれば全ての環境変数ファイルを誤ってコミットしなくなるか、変更を考えてみてください。

Karukan0814 commented on src/clothes_shop/models.py:
```diff
@@ -55,20 +57,46 @@ class Product(models.Model):
+    address = models.TextField(blank=True)
+    is_active = models.BooleanField(default=False) # 初回ログイン時に is_active を False に設定し、メール認証後に True にする
+    is_staff = models.BooleanField(default=False) # Djangoの管理画面にアクセスするための特定の権限を制御(Django管理画面を利用する場合は必須)
+    first_name = models.CharField(max_length=30, blank=True)
```
first_name, last_nameをいれた理由。フロント側の表示やら他テーブルにはねるので、必要なければ追加しないほうがよい。唯一許容は、ユーザー登録時、もしくはプロフィール画面からのユーザー名変更時にfirstname,lastnameわけて編集させる→バックエンド側で受け取ってfirstname,lastnameを合わせたものをusernameに登録。フロントやら他の処理では基本的にusernameを使用し、firstname,lastnameは登録・更新時のみ使用する項目にする。　それでも今回はtoo muchなやり方だから、追加する意味あんまないきがする。stripeの処理で必要なら必須の変更だが。。。　
理由次第なので要確認

Karukan0814 commented on src/clothes_shop/models.py:
```diff
@@ -55,20 +57,46 @@ class Product(models.Model):
+    password = models.CharField(max_length=128, null=True)  # パスワードフィールド追加
+    role = models.CharField(max_length=50, choices=[('guest', 'Guest'), ('registered', 'Registered'), ('admin', 'Admin')],default='registered')  # guest, registered, admin
+    address = models.TextField(blank=True)
+    is_active = models.BooleanField(default=False) # 初回ログイン時に is_active を False に設定し、メール認証後に True にする
```
guest={is_active=false,is_staff=false}  registered={is_active=true,is_staff=false}  admin={is_active=true,is_staff=true}で区分けができるのであれば、項目roleが必要ない。is_activeとis_staffがdjangoの標準モデルから継承したものであれば、roleはなおさら必要ない。こうした権限関係で管理に余分な項目を増やすのは論理エラーの元なので、role削除でいくべきだと思う。（※なお、これだけでもけっこうな修正なので、なおさらプルリクかコミット分けてコンセンサスとって事を進めるのが定石。）

だけれども、もはや手間が大変なので、次善の策としてmodelのrole項目をデフォルトでguestにしたうえで、論理エラーのないようにし、問題起きれば改めてブランチ切って対応ですね。

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,107 @@
+        # パスワードリセットトークンを生成し、メールを送信
+        try:
+            send_email(user, "reset")
+        except Exception as e:
```
エラーログを吐かせよう

Karukan0814 commented on src/clothes_shop/views.py:
```diff
@@ -188,13 +188,35 @@ class UserListCreateView(generics.ListCreateAPIView):
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
 
 class UserDetailView(generics.RetrieveUpdateDestroyAPIView):
```
clothes_shop側のテストフォルダにこの、viewとかserializerのテストコードがない

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,107 @@
+            return HttpResponse('無効なトークンです。', status=400)
+
+        if token_generator.check_token(user, token):
+            user.is_verified = True
```
is_verifiedっていう項目はあったっけ？モデルにないように見える。

Karukan0814 commented on src/djangopj/views.py:
```diff
@@ -0,0 +1,107 @@
+            return JsonResponse({"message": "無効なトークンです。"}, status=400)
+
+class UserListView(APIView):
+    def get(self, request):
```
このメソッドは何に使うのか？ユーザーのemailとパスワード、トークンとか重要情報を権限縛りもなしに提供している？

Karukan0814 commented on src/djangopj/urls.py:
```diff
@@ -1,7 +1,26 @@
+    path('api/auth/register/', RegisterUserView.as_view(), name='register'),
+    path('api/auth/logout/', LogoutView.as_view(), name='logout'),
+    path('api/auth/users/', UserListView.as_view(), name='user_list'),
+    path('api/auth/token/', TokenObtainPairView.as_view(), name='token_obtain_pair'),
```
これがログイン用？レスポンスでアクセストークンとリフレッシュトークンを返してくれる？

Karukan0814 commented on src/djangopj/urls.py:
```diff
@@ -1,7 +1,26 @@
+    path('api/auth/logout/', LogoutView.as_view(), name='logout'),
+    path('api/auth/users/', UserListView.as_view(), name='user_list'),
+    path('api/auth/token/', TokenObtainPairView.as_view(), name='token_obtain_pair'),
+    path('api/auth/token/refresh/', TokenRefreshView.as_view(), name='token_refresh'),
```
新しいアクセストークン取得用？

Karukan0814 commented on src/djangopj/urls.py:
```diff
@@ -1,7 +1,26 @@
+    path('api/auth/users/', UserListView.as_view(), name='user_list'),
+    path('api/auth/token/', TokenObtainPairView.as_view(), name='token_obtain_pair'),
+    path('api/auth/token/refresh/', TokenRefreshView.as_view(), name='token_refresh'),
+    path('api/auth/token/verify/', TokenVerifyView.as_view(), name='token_verify'),
```
トークンの検証用？

Karukan0814 commented:
[PR#19補足.md](https://github.com/user-attachments/files/17529962/PR.19.md)についてみましたが、バックエンドでトークンを差し込まなくても、現状の以下のデフォルト機能を使えばトークンをフロント側に返却してくれて、現状のフロントのjotaiストレージを若干拡張する形で流用可だと思うのですが、class LoginView(APIView):を作成してcookieに設定する利点がわからなかった。

from rest_framework_simplejwt.views import (
    TokenObtainPairView, TokenRefreshView, TokenVerifyView
)

Karukan0814 commented:
下記はバックエンド側でアクセストークンをヘッダーに組み込むイメージですがフロントエンド側で組み込む方針とする場合はまた別になる想定です
ログインビューを作成(今の構成なら clothes_shop/views.py 想定)し、ログイン時 Cookie にアクセストークン、リフレッシュトークンをセットする
バックエンド側で認証トークンハンドリングクラスを作成し(今の構成なら djangopj/authentication.py または clothes_shop/authentication.py に新規作成想定)、それを共通設定ファイル(今の構成なら djangopj/settings.py など想定)に設定する → 認可済みリクエストのみ許可する API(=registered ユーザーのみアクセス可能な API)へのリクエスト時の認証処理の過程で自動でリクエストヘッダーにトークンがセットできるようになる ログインビューの例(clothes_shop/views.py)
認証の結果、取得したトークンをクッキーに保存する



ここの部分がちょっと理解が追い付かなかった。ログイン機能をまだ作成していないor改修の必要ありという認識なのか？
シンプルに
ログインする→認証状態（有効なトークン持っている）→認証状態でしかリクエストできないAPIにアクセスできないようにする
をしたい場合、
・認証が必要なビュー：　permission_classes = [IsAuthenticated]　を使用
・それ以外：permission_classes = []　を使用
・フロント側：リクエスト飛ばす際にヘッダーにトークン含める。Authorization: `Bearer ${token}`,

で、認証可能だと考えていた。そして、現状のコードはどこまで実現できている想定なのか？ログイン認証できるの？

あと、adminかどうかはカスタムパーミッションクラスを作成する認識だが、合ってる？

Karukan0814 commented:
あと、まず認証関係の作成として、
1.ユーザーモデル決定
2.ユーザーシーダー作成して初期データ入れる
3.ログイン機能作成
4.認証されたリクエストしか特定のビューが使用できないように調整する方法確率
5.ユーザー登録の流れ

でやった方がいいと思います。特に認証機能なんて、根幹部分が3,4であり他の機能開発への影響も甚大。認証機能作成というタスクは3,4のこと。5は認証というよりユーザー登録です。
5のユーザー登録なんて最初はメール認証なんて後回しの単純なpostメソッドでいい。そこから拡張してメール機能組み入れ。

プルリク、もしくはコミットは最大この粒度、もっと細かくてもいいくらいだと思います。

Karukan0814 commented:
あと、User関係のテストコードがあんまないね。ビューだけじゃなく、モデル・シリアライザー、Urlのテストもあるよ。

sunny-side-app commented:
> PR#19補足.md についてみましたが、バックエンドでトークンを差し込まなくても、現状の以下のデフォルト機能を使えばトークンをフロント側に返却してくれて、現状のフロントのjotaiストレージを若干拡張する形で流用可だと思うのですが、class LoginView(APIView):を作成してcookieに設定する利点がわからなかった。
> 
> from rest_framework_simplejwt.views import (
> 
> TokenObtainPairView, TokenRefreshView, TokenVerifyView
> 
> )

フロント側のトークンの差し込み方の具体的なイメージまで検討しきれていなかったため、そちらとメリットデメリットを比較してこの実装案を提案したわけではないです。

「現状のフロントのjotaiストレージを若干拡張する形」というのは、useLogin.tsを基本的に使えるということなのかなと思ってますが、認識合ってますでしょうか。また、ご記載の方針の場合Cookieは認証関連目的で特に使う必要はない理解で合ってますか。

> ここの部分がちょっと理解が追い付かなかった。ログイン機能をまだ作成していないor改修の必要ありという認識なのか？
> シンプルに
> ログインする→認証状態（有効なトークン持っている）→認証状態でしかリクエストできないAPIにアクセスできないようにする
> をしたい場合、
> ・認証が必要なビュー：　permission_classes = [IsAuthenticated]　を使用
> ・それ以外：permission_classes = []　を使用
> ・フロント側：リクエスト飛ばす際にヘッダーにトークン含める。Authorization: `Bearer ${token}`,
> で、認証可能だと考えていた。そして、現状のコードはどこまで実現できている想定なのか？ログイン認証できるの？
> あと、adminかどうかはカスタムパーミッションクラスを作成する認識だが、合ってる？

↑の流れとシンプルに〜以降の記載も認識合ってます。

現状ログイン機能がどこまでできているかについては、↑の矢印の認証の流れを前提とすると、ユーザー情報(メールアドレス、パスワード等)をリクエストで渡してアクセストークンとリフレッシュトークンがレスポンスで返却されるエンドポイントを実装済みという意味で、最初の「ログインする」は出来ていると考えています(パスで言うと'api/auth/token/’が対応します)。

ログイン後のトークンの差し込み方：「・フロント側：リクエスト飛ばす際にヘッダーにトークン含める。Authorization: `Bearer ${token}`,」という部分を、ご記載のようにフロントでするのか、バックエンド側で自動でするようにするのか(補足資料に記載した方針)だけの認識齟齬なのかなと思ってまして、その差し込みを既存のuseLogin.tsでできるならば補足資料に記載のクラスコードは無視していただいて良いのかなと思ってます。

adminかどうかの制御については実装レベルの具体的なイメージはまだ湧いておりませんでした。自分で調査したり考えたいところですが時間がないため、もし参考になるような実装例などお持ちであれば共有いただけないでしょうか。


> あと、まず認証関係の作成として、
> 1.ユーザーモデル決定
> 2.ユーザーシーダー作成して初期データ入れる
> 3.ログイン機能作成
> 4.認証されたリクエストしか特定のビューが使用できないように調整する方法確率
> 5.ユーザー登録の流れ
> 
> でやった方がいいと思います。特に認証機能なんて、根幹部分が3,4であり他の機能開発への影響も甚大。認証機能作成というタスクは3,4のこと。5は認証というよりユーザー登録です。
> 5のユーザー登録なんて最初はメール認証なんて後回しの単純なpostメソッドでいい。そこから拡張してメール機能組み入れ。
> 
> プルリク、もしくはコミットは最大この粒度、もっと細かくてもいいくらいだと思います。
> 
> あと、User関係のテストコードがあんまないね。ビューだけじゃなく、モデル・シリアライザー、Urlのテストもあるよ。

流れについて承知しました。

1 のユーザーモデルについて確認させていただきたいですが、補足資料に記載のようにデフォルトのUserモデルを使うか、カスタムUserクラスを使うかでもしお考えなどあれば教えていただけますか。他のモデルオブジェクトと同じようにmodels.pyで扱えそうという意図で本PRではカスタムUserクラスにしていますが、設計上クラスを分けた意図はそれ以外ありません。

3 は、主にはバックエンド側のトークンを生成&リフレッシュするエンドポイント作成＋フロント側のuseLogin.tsを活用するような修正という、バックエンド側とフロントエンド側両方PRできて完成というイメージでしょうか。

4 については、adminはさておきですが、範囲は大きいかもしれませんが、認証が必要なビューにpermission_classes = [IsAuthenticated]　を使用しそれ以外：permission_classes = []　を使用するという実装を入れていくイメージですよね。

trelloに記載のタスク(ログイン、サインアップ、メール検証)の分割ができておらず、これを進めるべく認証系エンドポイントを一括で作成してしまいましたが、最低限コミットごとに1~5分けて進めるようにしたいと思います。

テストについては不足しており申し訳ありませんでした。追加したいと思います。

Karukan0814 reviewed (commented):
1 のユーザーモデルについて確認させていただきたいですが、補足資料に記載のようにデフォルトのUserモデルを使うか、カスタムUserクラスを使うかでもしお考えなどあれば教えていただけますか。他のモデルオブジェクトと同じようにmodels.pyで扱えそうという意図で本PRではカスタムUserクラスにしていますが、設計上クラスを分けた意図はそれ以外ありません。

→このままカスタムUserクラスでいいと思う。デフォルトUserクラスだと、ユーザーImgの項目やaddress、他にデフォルトで入っていない項目を入れたい場合、ちょっと困る気がする。このままカスタムUserクラスで認証問題なくできるなら、OK。あとは本当に必要な項目は何かを精査しましょう※他コメント参照。やっぱりroleの項目びみょーだなー。

3 は、主にはバックエンド側のトークンを生成&リフレッシュするエンドポイント作成＋フロント側のuseLogin.tsを活用するような修正という、バックエンド側とフロントエンド側両方PRできて完成というイメージでしょうか。

→Yes!　まずはバックエンド側を完成させる。フロントは最悪かるかんがまきとればいい。jotai関連の拡張だけで済むはず。

4 については、adminはさておきですが、範囲は大きいかもしれませんが、認証が必要なビューにpermission_classes = [IsAuthenticated]　を使用しそれ以外：permission_classes = []　を使用するという実装を入れていくイメージですよね。

→Yes!　実際に各APIに認証をいれていくのは各担当者がやればいいとは思うのですが、各担当者がその作業をできるように持っていく＝permission_classes = [IsAuthenticated]を入れれば認証が効く　にするのが目標です。

trelloに記載のタスク(ログイン、サインアップ、メール検証)の分割ができておらず、これを進めるべく認証系エンドポイントを一括で作成してしまいましたが、最低限コミットごとに1~5分けて進めるようにしたいと思います。

→コミットは細かくやらないと（※これは自分もあんまできてないが、、、）、何か戻りたいときに後悔するぜ。

Karukan0814 commented:
> > PR#19補足.md についてみましたが、バックエンドでトークンを差し込まなくても、現状の以下のデフォルト機能を使えばトークンをフロント側に返却してくれて、現状のフロントのjotaiストレージを若干拡張する形で流用可だと思うのですが、class LoginView(APIView):を作成してcookieに設定する利点がわからなかった。
> > from rest_framework_simplejwt.views import (
> > TokenObtainPairView, TokenRefreshView, TokenVerifyView
> > )
> 
> フロント側のトークンの差し込み方の具体的なイメージまで検討しきれていなかったため、そちらとメリットデメリットを比較してこの実装案を提案したわけではないです。
> 
> 「現状のフロントのjotaiストレージを若干拡張する形」というのは、useLogin.tsを基本的に使えるということなのかなと思ってますが、認識合ってますでしょうか。また、ご記載の方針の場合Cookieは認証関連目的で特に使う必要はない理解で合ってますか。
> 
> > ここの部分がちょっと理解が追い付かなかった。ログイン機能をまだ作成していないor改修の必要ありという認識なのか？
> > シンプルに
> > ログインする→認証状態（有効なトークン持っている）→認証状態でしかリクエストできないAPIにアクセスできないようにする
> > をしたい場合、
> > ・認証が必要なビュー：　permission_classes = [IsAuthenticated]　を使用
> > ・それ以外：permission_classes = []　を使用
> > ・フロント側：リクエスト飛ばす際にヘッダーにトークン含める。Authorization: `Bearer ${token}`,
> > で、認証可能だと考えていた。そして、現状のコードはどこまで実現できている想定なのか？ログイン認証できるの？
> > あと、adminかどうかはカスタムパーミッションクラスを作成する認識だが、合ってる？
> 
> ↑の流れとシンプルに〜以降の記載も認識合ってます。
> 
> 現状ログイン機能がどこまでできているかについては、↑の矢印の認証の流れを前提とすると、ユーザー情報(メールアドレス、パスワード等)をリクエストで渡してアクセストークンとリフレッシュトークンがレスポンスで返却されるエンドポイントを実装済みという意味で、最初の「ログインする」は出来ていると考えています(パスで言うと'api/auth/token/’が対応します)。
> 
> ログイン後のトークンの差し込み方：「・フロント側：リクエスト飛ばす際にヘッダーにトークン含める。Authorization: `Bearer ${token}`,」という部分を、ご記載のようにフロントでするのか、バックエンド側で自動でするようにするのか(補足資料に記載した方針)だけの認識齟齬なのかなと思ってまして、その差し込みを既存のuseLogin.tsでできるならば補足資料に記載のクラスコードは無視していただいて良いのかなと思ってます。
> 
> adminかどうかの制御については実装レベルの具体的なイメージはまだ湧いておりませんでした。自分で調査したり考えたいところですが時間がないため、もし参考になるような実装例などお持ちであれば共有いただけないでしょうか。
> 
> > あと、まず認証関係の作成として、
> > 1.ユーザーモデル決定
> > 2.ユーザーシーダー作成して初期データ入れる
> > 3.ログイン機能作成
> > 4.認証されたリクエストしか特定のビューが使用できないように調整する方法確率
> > 5.ユーザー登録の流れ
> > でやった方がいいと思います。特に認証機能なんて、根幹部分が3,4であり他の機能開発への影響も甚大。認証機能作成というタスクは3,4のこと。5は認証というよりユーザー登録です。
> > 5のユーザー登録なんて最初はメール認証なんて後回しの単純なpostメソッドでいい。そこから拡張してメール機能組み入れ。
> > プルリク、もしくはコミットは最大この粒度、もっと細かくてもいいくらいだと思います。
> > あと、User関係のテストコードがあんまないね。ビューだけじゃなく、モデル・シリアライザー、Urlのテストもあるよ。
> 
> 流れについて承知しました。
> 
> 1 のユーザーモデルについて確認させていただきたいですが、補足資料に記載のようにデフォルトのUserモデルを使うか、カスタムUserクラスを使うかでもしお考えなどあれば教えていただけますか。他のモデルオブジェクトと同じようにmodels.pyで扱えそうという意図で本PRではカスタムUserクラスにしていますが、設計上クラスを分けた意図はそれ以外ありません。
> 
> 3 は、主にはバックエンド側のトークンを生成&リフレッシュするエンドポイント作成＋フロント側のuseLogin.tsを活用するような修正という、バックエンド側とフロントエンド側両方PRできて完成というイメージでしょうか。
> 
> 4 については、adminはさておきですが、範囲は大きいかもしれませんが、認証が必要なビューにpermission_classes = [IsAuthenticated]　を使用しそれ以外：permission_classes = []　を使用するという実装を入れていくイメージですよね。
> 
> trelloに記載のタスク(ログイン、サインアップ、メール検証)の分割ができておらず、これを進めるべく認証系エンドポイントを一括で作成してしまいましたが、最低限コミットごとに1~5分けて進めるようにしたいと思います。
> 
> テストについては不足しており申し訳ありませんでした。追加したいと思います。



> 1 のユーザーモデルについて確認させていただきたいですが、補足資料に記載のようにデフォルトのUserモデルを使うか、カスタムUserクラスを使うかでもしお考えなどあれば教えていただけますか。他のモデルオブジェクトと同じようにmodels.pyで扱えそうという意図で本PRではカスタムUserクラスにしていますが、設計上クラスを分けた意図はそれ以外ありません。
> 
> →このままカスタムUserクラスでいいと思う。デフォルトUserクラスだと、ユーザーImgの項目やaddress、他にデフォルトで入っていない項目を入れたい場合、ちょっと困る気がする。このままカスタムUserクラスで認証問題なくできるなら、OK。あとは本当に必要な項目は何かを精査しましょう※他コメント参照。やっぱりroleの項目びみょーだなー。
> 
> 3 は、主にはバックエンド側のトークンを生成&リフレッシュするエンドポイント作成＋フロント側のuseLogin.tsを活用するような修正という、バックエンド側とフロントエンド側両方PRできて完成というイメージでしょうか。
> 
> →Yes!　まずはバックエンド側を完成させる。フロントは最悪かるかんがまきとればいい。jotai関連の拡張だけで済むはず。
> 
> 4 については、adminはさておきですが、範囲は大きいかもしれませんが、認証が必要なビューにpermission_classes = [IsAuthenticated]　を使用しそれ以外：permission_classes = []　を使用するという実装を入れていくイメージですよね。
> 
> →Yes!　実際に各APIに認証をいれていくのは各担当者がやればいいとは思うのですが、各担当者がその作業をできるように持っていく＝permission_classes = [IsAuthenticated]を入れれば認証が効く　にするのが目標です。
> 
> trelloに記載のタスク(ログイン、サインアップ、メール検証)の分割ができておらず、これを進めるべく認証系エンドポイントを一括で作成してしまいましたが、最低限コミットごとに1~5分けて進めるようにしたいと思います。
> 
> →コミットは細かくやらないと（※これは自分もあんまできてないが、、、）、何か戻りたいときに後悔するぜ。

Karukan0814 reviewed (commented):
みれる？

Karukan0814 committed:
add comment

Karukan0814 reviewed (changes requested):
このコメントみれる？

Karukan0814 reviewed (changes requested):
test
